java/src/com/ibm/streamsx/topology/generator/spl/SPLGenerator.java

```diff
@@ -83,6 +98,406 @@ public String generateSPL(JsonObject graph) throws IOException {
+        for(int i = 0; i < compStarts.size(); i++){
+            List<List<JsonObject> > startsEndsAndOperators = findCompositeOpsOfAType(graph, compStarts.get(i), compEnds.get(i), compOperatorStarts.get(i));
+
+            if(startsEndsAndOperators != null){
+                
+                JsonObject compDefinition;
+                // Composite definition contains list of operators
+                compDefinition = createCompositeDefinition(startsEndsAndOperators);
+                JsonObject compInvocation;
+
+                if(compStarts.get(i).equals(BVirtualMarker.PARALLEL.kind())){                  
+                    compInvocation = createParallelCompositeInvocation(compDefinition, startsEndsAndOperators);
+                }
+                else if(compStarts.get(i).equals(BVirtualMarker.LOW_LATENCY.kind())){
+                    compInvocation = createLowLatencyCompositeInvocation(compDefinition, startsEndsAndOperators);                  
+                }
+                else{
+                    throw new IllegalStateException("Unsupported composite type: " + compStarts.get(i));
+                } 
+                
+                // Fix the naming of the operators in the composite to read from the composite input ports
+                // and write to the composite output ports
+                fixCompositeInputNaming(graph, startsEndsAndOperators, compDefinition);
+                fixCompositeOutputNaming(graph, startsEndsAndOperators, compDefinition, compInvocation);
+                
+                // Add the invocation to the graph
+                array(graph, "operators").add(compInvocation);
+
+                // Remove starts, ends, and composite operators from the graph
+                for(int j = 0; j < startsEndsAndOperators.size(); j++){
+                    for(JsonElement op : startsEndsAndOperators.get(j)){
+                        array(graph, "operators").remove(op);
+                    }
+                }
+                
+                // Add the composite to the list of composites
+                stvHelper.addJsonParamDefs(compDefinition);
+                stvHelper.addJsonInstanceParams(compInvocation, compDefinition);
+                
+                composites.add(compDefinition);
+                
+                return true;
+            }
+        }
+        
+        return false;
+    }
+
+
+    /**
+     * When we create a composite, operators need to create connections with the composite's input port.
+     * @param graph
+     * @param startsEndsAndOperators
+     * @param opDefinition
+     */
+    private void fixCompositeInputNaming(JsonObject graph, List<List<JsonObject>> startsEndsAndOperators,
+            JsonObject opDefinition) {   
+        // For each start
+        // We iterate like this because we need to also index into the operatorDefinition's inputNames list.
+        for(int i = 0; i <  startsEndsAndOperators.get(0).size(); i++){
```

Comment:
Modern loops are much cleaner:
```
for(int i = 0; i <  startsEndsAndOperators.get(0).size(); i++){
    JsonObject start = startsEndsAndOperators.get(0).get(i);
```

can be:
```
for (JsonObject start : startsEndsAndOperators.get(0)) {

```

Reply:
But I need the counter to create a unique name for the input port. I could have a separate counter variable as well as doing the new style loop, but it seemed easier to just use a traditional loop.

Reply:
Which line of code is using `i` ? I just can't see it.

Reply:
You uncovered a bug.
```
connections.set(j, GsonUtilities.array(opDefinition, "inputNames").get(j));
```
should be 
```
connections.set(j, GsonUtilities.array(opDefinition, "inputNames").get(i));
```